natpmp/client.go

```diff
@@ -0,0 +1,204 @@
+package natpmp
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+// Version is the expected protocol version for NAT-PMP.
+const Version = 0
+
+// ErrProtocol indicates that a NAT gateway returned a response that violates
+// the NAT-PMP protocol.
+var ErrProtocol = errors.New("natpmp: protocol error")
+
+//go:generate stringer -type=Error -output=strings.go
+
+// An Error is a NAT-PMP protocol result code which indicates that an operation
+// has failed.
+type Error int
+
+// Possible Errors as defined in RFC 6886, section 3.5.
+const (
+	UnsupportedVersion Error = 1
+	NotAuthorized      Error = 2
+	NetworkFailure     Error = 3
+	OutOfResources     Error = 4
+	UnsupportedOpcode  Error = 5
+
+	// Although success is technically a result code, we don't expose it
+	// directly to the user because a successful operation returns nil error.
+	success Error = 0
+)
+
+// Error implements error.
+func (e Error) Error() string {
+	return fmt.Sprintf("natpmp: result %d: %s", e, e.String())
+}
+
+// A Client is a NAT-PMP client which can communicate with a NAT gateway.
+type Client struct {
+	// The UDP socket and address used to communicate with a NAT gateway.
+	mu      sync.Mutex
+	pc      net.PacketConn
+	gateway net.Addr
+}
+
+// Dial creates a Client which communicates with the NAT gateway specified by
+// addr.
+func Dial(addr string) (*Client, error) {
+	gateway, err := net.ResolveUDPAddr("udp4", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	pc, err := net.ListenPacket("udp4", ":0")
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(mdlayher): create a second listener on 224.0.0.1:5350 to listen
+	// to multicasts from a NAT gateway.
+
+	return &Client{
+		pc:      pc,
+		gateway: gateway,
+	}, nil
+}
+
+// Close closes the Client's underlying connection.
+func (c *Client) Close() error {
+	return c.pc.Close()
+}
+
+// An ExternalAddress is the result of a Client's ExternalAddress method.
+type ExternalAddress struct {
+	SinceStartOfEpoch time.Duration
+	ExternalIP        net.IP
```

Comment:
Any interest in using netaddr.IP here instead? It's a much better behaved value type, so you'd avoid retaining the entire packet in memory by returning a slice to it in ExternalAddress.

Reply:
I thought about it, but I hesitate to include netaddr in libraries which could be useful to the community as a whole rather than just a single project, especially because we haven't provided any stability guarantees there yet.

Reply:
Admittedly my personal bias, having used netaddr extensively in Tailscale, is that it's so much better (more performant, less error-prone, ...) than net's types that it's worth requiring it. At this point I consider net.IP to be a legacy interface, and netaddr is what should have been in stdlib. So, I would definitely advocate for us using netaddr in this API.

That said, as long as this package doesn't hit 1.0, we can argue over that separately :)